sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeter.java

```diff
@@ -99,7 +99,7 @@ public BatchRecorderSdk newBatchRecorder(String... keyValuePairs) {
     Collection<AbstractInstrument> instruments = instrumentRegistry.getInstruments();
     List<MetricData> result = new ArrayList<>(instruments.size());
     for (AbstractInstrument instrument : instruments) {
-      result.addAll(instrument.collectAll());
+      instrument.collectAll(result);
```

Comment:
As this is an internal method and there does seem to be significant allocation churn here I suspect we will want to make the change. But I agree that we should always provide the before / after jmh for performance improvements.

Reply:
For reference, rewrite proto serialization had a much larger negative effect on maintainability :) But we think it's worth it thanks to the numbers.

Reply:
One list per instrument per collection cycle doesn't feel like much to me.

Reply:
I may be wrong but isn't an instrument what actually measures a given metric? I guess apps often have hundreds of them which seems like a point that could use some tweaks.

Reply:
Compared to the allocations with everything else involved in creating measurements, I doubt this would have much impact, but to re-iterate, we should measure the difference so we have data to base a decision on.